packages/libp2p/test/core/random-walk.spec.ts

```diff
@@ -0,0 +1,169 @@
+    let yielded = 0
+
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * (key, options?: AbortOptions) {
+        for (let i = 0; i < 100; i++) {
+          yielded++
+          yield await createRandomPeerInfo()
+          options?.signal?.throwIfAborted()
+        }
+      }())
+      .onSecondCall().returns(slowIterator())
+
+    await drain(take(randomwalk.walk(), 1))
+
+    expect(yielded).to.equal(1)
+  })
+
+  it('should keep walking until done', async () => {
+    const randomPeer1 = await createRandomPeerInfo()
+    const randomPeer2 = await createRandomPeerInfo()
+
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * () {
+        yield randomPeer1
+      }())
+      .onSecondCall().returns(async function * () {
+        yield randomPeer2
+      }())
+      .onThirdCall().returns(slowIterator())
+
+    const peers = await all(take(randomwalk.walk(), 2))
+
+    expect(peers.map(peer => peer.id.toString())).to.deep.equal([
+      randomPeer1.id.toString(),
+      randomPeer2.id.toString()
+    ])
+  })
+
+  it('should join an existing random walk', async () => {
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * () {
+        for (let i = 0; i < 100; i++) {
+          yield await createRandomPeerInfo()
+          await delay(100)
+        }
+      }())
+      .onSecondCall().returns(slowIterator())
+
+    const [
+      peers1,
+      peers2
+    ] = await Promise.all([
+      all(take(randomwalk.walk(), 2)),
+      all(take(randomwalk.walk(), 2))
+    ])
+
+    expect(peers1.map(peer => peer.id.toString())).to.deep.equal(peers2.map(peer => peer.id.toString()))
+  })
+
+  it('should continue random walk until all consumers satisfied', async () => {
+    let yielded = 0
+
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * (key, options?: AbortOptions) {
+        for (let i = 0; i < 100; i++) {
+          yielded++
+          yield await createRandomPeerInfo()
+          options?.signal?.throwIfAborted()
+        }
+      }())
+      .onSecondCall().returns(slowIterator())
+
+    await Promise.all([
+      drain(take(randomwalk.walk(), 1)),
+      drain(take(randomwalk.walk(), 2))
+    ])
+
+    expect(yielded).to.equal(2)
+  })
+
+  it('should not block walk on slow consumers', async () => {
+    let yielded = 0
+
+    peerRouting.getClosestPeers
+      .onFirstCall().returns(async function * (key, options?: AbortOptions) {
+        for (let i = 0; i < 100; i++) {
+          yielded++
+          yield await createRandomPeerInfo()
+          options?.signal?.throwIfAborted()
+        }
+      }())
+      .onSecondCall().returns(slowIterator())
+
+    await Promise.all([
+      drain(take(randomwalk.walk(), 5)),
+      drain(map(take(randomwalk.walk(), 2), async peer => {
+        await delay(100)
+        return peer
+      }))
+    ])
+
+    expect(yielded).to.equal(7)
+  })
```

Comment:
this test name implies that we're confirming that take5 (line 160) resolves faster than take2-slow (line 161), but we're only confirming the total?

Should we confirm ordering somehow? maybe with a .then() and timestamp?

Reply:
The name says slow consumers don't block the walk, so the assertion asserts that the walk continued despite one consumer being slow.